core-bundle/src/Framework/ContaoFramework.php

```diff
@@ -430,4 +441,16 @@ private function registerHookListeners(): void
             $GLOBALS['TL_HOOKS'][$hookName] = array_merge(...$priorities);
         }
     }
+
+    private function throwOnLegacyRoutingHooks(): void
+    {
+        if (
+            empty($GLOBALS['TL_HOOKS']['getPageIdFromUrl'])
+            && empty($GLOBALS['TL_HOOKS']['getRootPageFromUrl'])
+        ) {
+            return;
+        }
+
+        throw new LegacyRoutingException('Legacy routing is required to support the getPageIdFromUrl and getRootPageFromUrl hooks. Check the Symfony inspector for more information.');
```

Comment:
This means that extensions that try to work for both versions need to dynamically register the hook, right?
And while we're on it: The message indicates that you should enable legacy routing but I think what you actually want is to inform the extension developer about the changes required :)
What about something more specific like

> Some extension services (<comma-separated-class-or-service-names>) are trying to use the deprecated "getPageIdFromUrl" or "getRootPageFromURL" hooks. This is incompatible with your current routing setup. You either have to enable the legacy routing mode by ..... to support these extensions or you have to uninstall/update them to a version compatible with the new routing mode.

Reply:
I agree the message could be more descriptive. But I wonder if the exception message is the right place for this. Maybe we should – after all – add another exception handling to generate a nice error page? **Or** this could be added to the documentation and the exception message should include the link?

Reply:
I think the error message should be more descriptive. Simple and done.

Reply:
This is still open.